File: ImperativeParadigm/Exercise-1/Go/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)

// File represents a file with its name and content
type File struct {
    filename string
    content string
}

// Metadata represents the metadata of a file
type Metadata struct {
    file File
    lines int
    words int 
    characters int
}

// readFile reads the content of a file 
//
// It returns an error if the file cannot be read
// It returns nil if the file is read successfully
func (f *File) readFile() error {
    content, err := os.ReadFile(f.filename)
    if err != nil {
        return err
    }

    f.content = string(content)
    return nil
}

// countLines counts the number of lines in a file
func (m *Metadata) countLines() {
    m.lines = len(strings.Split(m.file.content, "\n"))
}

// countWords counts the number of words in a file
func (m *Metadata) countWords() {
    m.words = len(strings.Fields(m.file.content))
}

// countCharacters counts the number of characters in a file
func (m *Metadata) countCharacters() {
    m.characters = len(m.file.content)
}

// collect collects the metadata of a file
func (m *Metadata) collect() {
    m.countLines()
    m.countWords()
    m.countCharacters()
}

// toString converts the metadata of a file to a string
func (m *Metadata) toString() string {
    return fmt.Sprintf("File: %s\nLines: %d\nWords: %d\nCharacters: %d\n", m.file.filename, m.lines, m.words, m.characters)
}

// This program reads the content of a file and prints its metadata
func main() {
    file := File{filename: "lorem-tiny.txt"}
    err := file.readFile()
    if err != nil {
        panic(err)
    }

    metadata := Metadata{file: file}
    metadata.collect()
    fmt.Println(metadata.toString())
}

File: ImperativeParadigm/Exercise-1/Rust/src/main.rs
struct File {
    name: String,
    data: Vec<u8>,
}

impl File {
    fn new(name: &str) -> File {
        File {
            name: String::from(name),
            data: Vec::new(),
        }
    }

    fn read(&mut self) {
        self.data = std::fs::read(&self.name).unwrap();
    }
}

struct Metadata {
    file: File,
    lines: u32,
    words: u32,
    chars: u32,
}

impl Metadata {
    fn new(file: File) -> Metadata {
        Metadata {
            file,
            lines: 0,
            words: 0,
            chars: 0,
        }
    }

    fn count_lines(&mut self) {
        let mut lines = 0;
        for byte in &self.file.data {
            if *byte == b'\n' {
                lines += 1;
            }
        }
        self.lines = lines;
    }

    fn count_words(&mut self) {
        let mut words = 0;
        let mut in_word = false;
        for byte in &self.file.data {
            match *byte {
                b'\n' | b'\t' | b' ' => {
                    if in_word {
                        in_word = false;
                        words += 1;
                    }
                }
                _ => in_word = true,
            }
        }
        self.words = words;
    }

    fn count_chars(&mut self) {
        self.chars = self.file.data.len() as u32;
    }

    fn collect(&mut self) {
        self.count_lines();
        self.count_words();
        self.count_chars();
    }
}

fn main() {
    let mut file = File::new("lorem-tiny.txt");
    file.read();

    let mut metadata = Metadata::new(file);
    metadata.collect();

    println!(
        "lines: {}, words: {}, chars: {}",
        metadata.lines, metadata.words, metadata.chars
    );
}
